src/server/frontend_wayland/idle_inhibit_v1.cpp

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright © 2022 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: Grayson Guarino <grayson.guarino@canonical.com>
+ */
+
+#include "idle_inhibit_v1.h"
+
+#include "mir/log.h"
+#include "mir/executor.h"
+#include "mir/scene/surface.h"
+#include "mir/scene/null_surface_observer.h"
+
+#include "wl_surface.h"
+
+#include <boost/throw_exception.hpp>
+#include <memory>
+
+namespace mf = mir::frontend;
+namespace mw = mir::wayland;
+namespace ms = mir::scene;
+
+namespace mir
+{
+namespace frontend
+{
+struct IdleInhibitV1Ctx
+{
+    class SurfaceObserver : ms::NullSurfaceObserver
+    {
+    public:
+        SurfaceObserver()
+        {
+        }
+
+        ~SurfaceObserver()
+        {
+        }
+
+        void attrib_changed(const ms::Surface *surf, MirWindowAttrib attrib, int value) override
+        {
+            (void)surf;
+            if (attrib == mir_window_attrib_focus && value == mir_window_focus_state_unfocused)
+                ; // TODO
+            else if (attrib == mir_window_attrib_focus)
+                ; // TODO
+        };
+    };
+
+    std::shared_ptr<Executor> const wayland_executor;
+    std::shared_ptr<ms::IdleHub> const idle_hub;
+    std::shared_ptr<ms::SurfaceObserver> surface_observer;
```

Comment:
I'm not a big fan of shared structs with mutable state: it can become hard to enforce invariants.

Relatedly I also see `ctx->idle_hub->inhibit_idle()` which might better be written `ctx->inhibit_idle()` if we made this a class (with member functions and private data).

Reply:
The point of the ctx object is to provide access to a set of other things with as little code overhead as possible. In my opinion it's not supposed to encapsulate any functionality.

I agree it's not the right place for mutable state.

Reply:
Upon review, there is no reason that `surface_observer` needs to be mutable. I believe that got held over from my previous implementation.

If the declaration of `SurfaceObserver` exists within `IdleInhibitV1Ctx`, *without* a mutable state of `surface_observer`, should I still make this a class or leave it as a struct.

Reply:
> Alan may disagree.

In general, I'd prefer to keep state directly in an object rather than in a shared subobject. But my primary concern here was _shared mutable state_.

Specifically, I'd either:

1. Copy the `wayland_executor`, `idle_hub` and `surface_observer` from the `ctx` supplied to the constructor as required and use them directly; or,
2. Make `ctx` into a class that forwards requests to the corresponding subobjects.